airflow/models/dag_version.py

```diff
@@ -0,0 +1,163 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from __future__ import annotations
+
+import logging
+import random
+import string
+from typing import TYPE_CHECKING
+
+import uuid6
+from sqlalchemy import Column, ForeignKey, Integer, UniqueConstraint, func, select
+from sqlalchemy.orm import relationship
+from sqlalchemy_utils import UUIDType
+
+from airflow.models.base import Base, StringID
+from airflow.utils import timezone
+from airflow.utils.session import NEW_SESSION, provide_session
+from airflow.utils.sqlalchemy import UtcDateTime, with_row_locks
+
+if TYPE_CHECKING:
+    from sqlalchemy.orm import Session
+
+
+log = logging.getLogger(__name__)
+
+
+def _gen_random_str():
+    return "".join(random.choices(string.ascii_letters + string.digits, k=10))
+
+
+class DagVersion(Base):
+    """Model to track the versions of DAGs in the database."""
+
+    __tablename__ = "dag_version"
+    id = Column(UUIDType(binary=False), primary_key=True, default=uuid6.uuid7)
+    version_number = Column(Integer, nullable=False, default=1)
+    version_name = Column(StringID(), default=_gen_random_str, nullable=False)
+    dag_id = Column(StringID(), ForeignKey("dag.dag_id", ondelete="CASCADE"), nullable=False)
+    dag_model = relationship("DagModel", back_populates="dag_versions")
+    dag_code = relationship(
+        "DagCode",
+        back_populates="dag_version",
+        uselist=False,
+        cascade="all, delete, delete-orphan",
+        cascade_backrefs=False,
+    )
+    serialized_dag = relationship(
+        "SerializedDagModel",
+        back_populates="dag_version",
+        uselist=False,
+        cascade="all, delete, delete-orphan",
+        cascade_backrefs=False,
+    )
+    dag_runs = relationship("DagRun", back_populates="dag_version", cascade="all, delete, delete-orphan")
+    task_instances = relationship("TaskInstance", back_populates="dag_version")
+    created_at = Column(UtcDateTime, default=timezone.utcnow)
+
+    __table_args__ = (
+        UniqueConstraint("dag_id", "version_number", name="dag_id_version_number_unique_constraint"),
+    )
+
+    def __repr__(self):
+        return f"<DagVersion {self.dag_id} - {self.version_name}>"
```

Comment:
```suggestion
        return f"<DagVersion {self.dag_id} {self.version}>"
```

Aren't we missing half of the logical key here?